src/data/mod.rs

```diff
@@ -468,46 +469,112 @@ where
     temp.into()
 }
 
-macro_rules! data_func_def {
-    ($doc_str: expr, $fn_name:ident, $ffi_name: ident) => {
-        #[doc=$doc_str]
-        ///
-        ///# Parameters
-        ///
-        /// - `input` is the input Array
-        /// - `dims` is the target(output) dimensions
-        ///
-        ///# Return Values
-        ///
-        /// An Array with modified data.
-        #[allow(unused_mut)]
-        pub fn $fn_name<T>(input: &Array<T>, dims: Dim4) -> Array<T>
-        where
-            T: HasAfEnum,
-        {
-            let mut temp: i64 = 0;
-            unsafe {
-                let err_val = $ffi_name(
-                    &mut temp as MutAfArray,
-                    input.get() as AfArray,
-                    dims[0] as c_uint,
-                    dims[1] as c_uint,
-                    dims[2] as c_uint,
-                    dims[3] as c_uint,
-                );
-                HANDLE_ERROR(AfError::from(err_val));
+/// Tile the input array along specified dimension
+///
+/// Tile essentially creates copies of data along each dimension.
+/// The number of copies created is provided by the user on per
+/// axis basis using [Dim4](./struct.dim4.html)
+///
+///# Parameters
+///
+/// - `input` is the input Array
+/// - `dims` is the target(output) dimensions
+///
+///# Return Values
+///
+/// Tiled input array as per the tiling dimensions provided
+#[allow(unused_mut)]
+pub fn tile<T>(input: &Array<T>, dims: Dim4) -> Array<T>
+where
+    T: HasAfEnum,
+{
+    let mut temp: i64 = 0;
+    unsafe {
+        let err_val = af_tile(
+            &mut temp as MutAfArray,
+            input.get() as AfArray,
+            dims[0] as c_uint,
+            dims[1] as c_uint,
+            dims[2] as c_uint,
+            dims[3] as c_uint,
+        );
+        HANDLE_ERROR(AfError::from(err_val));
+    }
+    temp.into()
+}
+
+/// Reorder the array in specified order
+///
+/// The default order of axes in ArrayFire is axis with smallest distance
+/// between adjacent elements towards an axis with highest distance between
+/// adjacent elements.
+///
+///# Parameters
+///
+/// - `input` is the input Array
+/// - `next_axes` is the new axeses order for output
+///
+///# Return Values
+///
+/// Array with data reordered as per the new axes order
+pub fn reorder_v2<T>(
```

Comment:
I would suggest `reorder<T>(input: &Array<T>, new_axes: Vec<u64>)` plain and simple. That way if new dimensions are added to ArrayFire you won't have to update this function. The signature that you propose could be added as a swap function such as `swap_axes<T>(input: &Array<T>, new_axis0: u64, new_axis1: u64)` in a future release for instance.

Reply:
`reorder` always requires at least two arguments - a minimum of two axes have to be provided to successfully swap data between two axes. Passing a single argument is invalid input and will only result in exception from ArrayFire upstream. Therefore, by enforcing two mandatory arguments for `reorder_v2` I am avoiding an unnecessary round trip to FFI land from rust, which also accurately reflects the functionality of reorder. Hence, the optional list of arguments (vec) is wrapped in `std::opt::Option`. For the prior mentioned reasons, I chose two mandatory u64 arguments followed by an optional vector which will cover all cases without any future modifications. We don't need another function like swap_axes at all.